crates/core/src/fs.rs

```diff
@@ -1,51 +1,64 @@
 use opendal::raw::{build_rooted_abs_path, normalize_path, normalize_root};
 use opendal::{EntryMode, ErrorKind, Operator};
-use pyo3::exceptions::{PyException, PyFileNotFoundError};
+use pyo3::exceptions::{PyException, PyFileNotFoundError, PyValueError};
 use pyo3::prelude::*;
-use pyo3::types::{PyDict, PyList};
+use pyo3::types::{PyDateTime, PyDict};
+use tokio::runtime::Runtime;
 
 #[pyclass(subclass)]
 pub struct OpendalFileSystem {
     op: Operator,
+    rt: Runtime,
 }
 
 impl From<Operator> for OpendalFileSystem {
     fn from(op: Operator) -> Self {
-        Self { op }
+        Self {
+            op,
+            rt: Runtime::new().unwrap(),
+        }
     }
 }
 
 #[pymethods]
 impl OpendalFileSystem {
-    fn ls<'py>(&self, py: Python<'py>, path: &str) -> PyResult<pyo3::Bound<'py, PyList>> {
-        let entries = self
-            .op
-            .blocking()
-            .list(path)
-            .unwrap()
-            .into_iter()
-            .map(|v| v.path().to_string())
-            .collect::<Vec<_>>();
-        Ok(PyList::new_bound(py, &entries))
+    /// List contents of a path
+    fn ls(&self, _py: Python, path: &str) -> PyResult<Vec<String>> {
+        let path = normalize_path(path);
+
+        // Use runtime to execute async list for S3
+        self.rt.block_on(async {
```

Comment:
This is nearly the same thing we did in opendal's blocking layer, we can enable it by:

https://github.com/apache/opendal/blob/52c96bb8e8cb4d024ccab1f415c4756447c726dd/bindings/python/src/operator.rs#L35-L43